src/main/java/com/google/sps/servlets/EditBusinessDataServlet.java

```diff
@@ -0,0 +1,248 @@
+    
+    businessEntity.setProperty(BUSINESS_NAME, name);
+    businessEntity.setProperty(BUSINESS_DESC, desc);
+    businessEntity.setProperty(BUSINESS_CATEGORIES, gson.toJson(categories));
+    businessEntity.setProperty(BUSINESS_ADDRESS, address);
+    businessEntity.setProperty(BUSINESS_ADDRESS_LAT, addressLat);
+    businessEntity.setProperty(BUSINESS_ADDRESS_LNG, addressLng);
+    businessEntity.setProperty(BUSINESS_CONTACT_INFO, contactDetails);
+    businessEntity.setProperty(BUSINESS_ORDER_INFO, orderDetails);
+    businessEntity.setProperty(BUSINESS_LINK, businessLink);
+    businessEntity.setProperty(BUSINESS_MENU_LINK, menuLink);
+    businessEntity.setProperty(BUSINESS_MIN_PRICE, minPrice);
+    businessEntity.setProperty(BUSINESS_MAX_PRICE, maxPrice);
+    businessEntity.setProperty(BUSINESS_LOGO, logoUrl);
+    businessEntity.setProperty(BUSINESS_PICTURES, gson.toJson(picturesUrls));
+    businessEntity.setProperty(BUSINESS_RATING, rating);
+    businessEntity.setProperty(BUSINESS_REVIEWS, reviews);
+
+    storeBusinessEntity(datastore, businessEntity);
+
+    // Redirect back to the product page page.
+    response.sendRedirect("/index.html");
+    } catch (IOException err) {
+      System.out.println(err);
+    } catch (EntityNotFoundException err) {
+      System.out.println(err);
+    } 
+  }
+
+  private Entity getBusinessEntity(DatastoreService datastore, Key businessKey) throws EntityNotFoundException {
+    Entity businessEntity = datastore.get(businessKey);
+    return businessEntity;
+  }
+
+
+  private void storeBusinessEntity(DatastoreService datastore, Entity businessEntity) throws EntityNotFoundException {
+    datastore.put(businessEntity);
+  }
+
+  private String getUploadedLogoUrlFromBlobstore(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // if no images were uploaded
+    if (blobKeys == null) {
+        return "";
+    }
+
+    // Our form only contains a single file input, so get the first index.
+    BlobKey blobKey = blobKeys.get(0);
+
+    // We could check the validity of the file here, e.g. to make sure it's an image file
+    // https://stackoverflow.com/q/10779564/873165
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+    String url = imagesService.getServingUrl(options);
+
+    // GCS's localhost preview is not actually on localhost,
+    // so make the URL relative to the current domain.
+    if(url.startsWith("http://localhost:8080/")){
+      url = url.replace("http://localhost:8080/", "/");
+    }
+    return url;
```

Comment:
I'm nitpicking, but consider refactorizing this portion to a helper function, and reusing it in both functions.

Reply:
Actually, would writing a single function (taking a BlobKey and returning a url) make sense here? So we use only that function and for the List<String> of product photos, we loop over the blobkeys? I feel like there's repeated code across both functions. Up for discussion.

Reply:
yup, you're right! I've kept only the second function that returns a list of string. for the logo i just take the first (and only string) in that list

Reply:
i've done the same in my BusinessDataServlet.java form